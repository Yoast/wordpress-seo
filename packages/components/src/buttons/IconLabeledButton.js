// External dependencies.
import React from "react";
import styled from "styled-components";
import PropTypes from "prop-types";
import flow from "lodash/flow";
import omit from "lodash/omit";

// Yoast dependencies.
import { colors } from "@yoast/style-guide";
import SvgIcon from "../SvgIcon";

// Internal dependencies.
import { addActiveStyle, addFocusStyle, addHoverStyle } from "./Button";

/**
 * Returns a component with all button selector styles applied.
 *
 * @param {ReactElement} component The original component.
 *
 * @returns {ReactElement} Component with applied styles.
 */
const addButtonStyles = flow( [
	/*
	 * Styled-components applies the generated CSS classes in a reversed order,
	 * but we want them in the order: base - hover - focus - active.
	 */
	addActiveStyle,
	addFocusStyle,
	addHoverStyle,
] );

/**
 * Returns a basic styled button.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled button.
 */
const IconLabelledBaseButton = addButtonStyles(
	styled.button`
		display: inline-flex;
		flex-direction: column;
		justify-content: center;
		align-items: center;
		cursor: pointer;
		box-sizing: border-box;
		border: 1px solid transparent;
		margin: 0;
		padding: 8px;
		overflow: visible;
		font-family: inherit;
		font-weight: inherit;
		color: ${ props => props.textColor };
		background: ${ props => props.backgroundColor };
		font-size: ${ props => props.textFontSize };

		svg {
			margin: 0 0 8px;
			flex-shrink: 0;
			fill: currentColor;
			// Safari 10
			align-self: center;
		}

		&:active {
			box-shadow: none;
		}
	`
);

/**
 * Returns an icon button that can optionally contain text.
 *
 * @param {object} props Component props.
 *
 * @returns {ReactElement} Styled icon button.
 */
const IconLabeledButton = ( props ) => {
	const { children, icon, textColor } = props;

	const newProps = omit( props, "icon" );

	return (
		<IconLabelledBaseButton { ...newProps }>
			<SvgIcon icon={ icon } color={ textColor } />
			{ children }
		</IconLabelledBaseButton>
	);
};

IconLabeledButton.propTypes = {
	type: PropTypes.string,
	icon: PropTypes.string.isRequired,
	textColor: PropTypes.string,
	textFontSize: PropTypes.string,
	backgroundColor: PropTypes.string,
	borderColor: PropTypes.string,
	hoverColor: PropTypes.string,
	hoverBackgroundColor: PropTypes.string,
	hoverBorderColor: PropTypes.string,
	activeColor: PropTypes.string,
	activeBackgroundColor: PropTypes.string,
	activeBorderColor: PropTypes.string,
	focusColor: PropTypes.string,
	focusBackgroundColor: PropTypes.string,
	focusBorderColor: PropTypes.string,
	focusBoxShadowColor: PropTypes.string,
	children: PropTypes.oneOfType( [
		PropTypes.arrayOf( PropTypes.node ),
		PropTypes.node,
		PropTypes.string,
	] ).isRequired,
};

IconLabeledButton.defaultProps = {
	type: "button",
	textColor: colors.$color_blue,
	textFontSize: "inherit",
	backgroundColor: "transparent",
	borderColor: "transparent",
	hoverColor: colors.$color_white,
	hoverBackgroundColor: colors.$color_blue,
	hoverBorderColor: colors.$color_button_border_hover,
	activeColor: colors.$color_white,
	activeBackgroundColor: colors.$color_blue,
	activeBorderColor: colors.$color_button_border_active,
	focusColor: colors.$color_white,
	focusBackgroundColor: colors.$color_blue,
	focusBorderColor: colors.$color_blue,
	focusBoxShadowColor: colors.$color_blue_dark,
};

export default IconLabeledButton;
