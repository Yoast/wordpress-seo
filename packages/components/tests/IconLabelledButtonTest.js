import React from "react";
import renderer from "react-test-renderer";

import IconLabeledButton from "../src/buttons/IconLabeledButton";

test( "the IconLabelledButton matches the snapshot", () => {
	const component = renderer.create(
		<IconLabeledButton icon="question-circle">Need help?</IconLabeledButton>
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );

test( "the IconLabelledButton with props matches the snapshot", () => {
	const component = renderer.create(
		<IconLabeledButton
			type="reset"
			icon="eye"
			iconColor="#0f0"
			textColor="#0f0"
			textFontSize="1em"
			backgroundColor="#ff0"
			borderColor="#000"
			hoverColor="#fff"
			hoverBackgroundColor="#a4286a"
			hoverBorderColor="#f00"
			activeColor="#fff"
			activeBackgroundColor="#a4286a"
			activeBorderColor="#a4286a"
			focusColor="#a4286a"
			focusBackgroundColor="#e1bee7"
			focusBorderColor="#a4286a"
		>
			Hello
		</IconLabeledButton>
	);

	const tree = component.toJSON();
	expect( tree ).toMatchSnapshot();
} );
