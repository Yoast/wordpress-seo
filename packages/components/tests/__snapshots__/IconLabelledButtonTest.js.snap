// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the IconLabelledButton matches the snapshot 1`] = `
.c4 {
  width: 16px;
  height: 16px;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.c1:active {
  color: #fff;
  background-color: #0066cd;
  border-color: #888;
  box-shadow: inset 0 2px 5px -3px rgba( 0,0,0,0.5 );
}

.c2::-moz-focus-inner {
  border-width: 0;
}

.c2:focus {
  outline: none;
  border-color: #0066cd;
  color: #fff;
  background-color: #0066cd;
  box-shadow: 0 0 3px rgba( 8,74,103,0.8 );
}

.c3:hover {
  color: #fff;
  background-color: #0066cd;
  border-color: var(--yoast-color-border--default);
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  box-sizing: border-box;
  border: 1px solid transparent;
  margin: 0;
  padding: 8px;
  overflow: visible;
  font-family: inherit;
  font-weight: inherit;
  color: #0066cd;
  background: transparent;
  font-size: inherit;
}

.c0 svg {
  margin: 0 0 8px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  fill: currentColor;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c0:active {
  box-shadow: none;
}

<button
  className="c0 c1 c2 c3"
  type="button"
>
  <svg
    aria-hidden={true}
    className="c4 yoast-svg-icon yoast-svg-icon-question-circle"
    fill="#0066cd"
    focusable="false"
    role="img"
    size="16px"
    viewBox="0 0 1792 1792"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M1024 1376v-192q0-14-9-23t-23-9h-192q-14 0-23 9t-9 23v192q0 14 9 23t23 9h192q14 0 23-9t9-23zm256-672q0-88-55.5-163t-138.5-116-170-41q-243 0-371 213-15 24 8 42l132 100q7 6 19 6 16 0 25-12 53-68 86-92 34-24 86-24 48 0 85.5 26t37.5 59q0 38-20 61t-68 45q-63 28-115.5 86.5t-52.5 125.5v36q0 14 9 23t23 9h192q14 0 23-9t9-23q0-19 21.5-49.5t54.5-49.5q32-18 49-28.5t46-35 44.5-48 28-60.5 12.5-81zm384 192q0 209-103 385.5t-279.5 279.5-385.5 103-385.5-103-279.5-279.5-103-385.5 103-385.5 279.5-279.5 385.5-103 385.5 103 279.5 279.5 103 385.5z"
    />
  </svg>
  Need help?
</button>
`;

exports[`the IconLabelledButton with props matches the snapshot 1`] = `
.c4 {
  width: 16px;
  height: 16px;
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.c1:active {
  color: #fff;
  background-color: #a4286a;
  border-color: #f00;
  box-shadow: inset 0 2px 5px -3px rgba( 164,40,106,0.5 );
}

.c2::-moz-focus-inner {
  border-width: 0;
}

.c2:focus {
  outline: none;
  border-color: #a4286a;
  color: #a4286a;
  background-color: #e1bee7;
  box-shadow: 0 0 3px rgba( 8,74,103,0.8 );
}

.c3:hover {
  color: #fff;
  background-color: #a4286a;
  border-color: var(--yoast-color-border--default);
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  box-sizing: border-box;
  border: 1px solid transparent;
  margin: 0;
  padding: 8px;
  overflow: visible;
  font-family: inherit;
  font-weight: inherit;
  color: #0f0;
  background: #ff0;
  font-size: 1em;
}

.c0 svg {
  margin: 0 0 8px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  fill: currentColor;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

.c0:active {
  box-shadow: none;
}

<button
  className="c0 c1 c2 c3"
  type="reset"
>
  <svg
    aria-hidden={true}
    className="c4 yoast-svg-icon yoast-svg-icon-eye"
    fill="#0f0"
    focusable="false"
    role="img"
    size="16px"
    viewBox="0 0 1792 1792"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M1664 960q-152-236-381-353 61 104 61 225 0 185-131.5 316.5t-316.5 131.5-316.5-131.5-131.5-316.5q0-121 61-225-229 117-381 353 133 205 333.5 326.5t434.5 121.5 434.5-121.5 333.5-326.5zm-720-384q0-20-14-34t-34-14q-125 0-214.5 89.5t-89.5 214.5q0 20 14 34t34 14 34-14 14-34q0-86 61-147t147-61q20 0 34-14t14-34zm848 384q0 34-20 69-140 230-376.5 368.5t-499.5 138.5-499.5-139-376.5-368q-20-35-20-69t20-69q140-229 376.5-368t499.5-139 499.5 139 376.5 368q20 35 20 69z"
    />
  </svg>
  Hello
</button>
`;
